bulletroboy/operator.py

```diff
@@ -0,0 +1,203 @@
+import pybullet as p
+import time
+import math
+import numpy as np
+from numpy.linalg import norm
+from termcolor import colored
+
+
+class Operator():
+	def __init__(self, body_id):
+		"""
+		This class handles the operator body and its links in the simulation.
+		"""
+		self.body_id = body_id
+		self.links = self.get_links()
+		self.movements = Movements(self)
+
+	def get_links(self):
+		links = []
+		for i in range(p.getNumJoints(self.body_id)):
+			link = {}
+			link['name'] = str(p.getJointInfo(self.body_id,i)[12], 'utf-8')
+			link['id'] = i
+			links.append(link)
+		return links
+
+	def get_link_center(self, link_name):
+		center = None
+		index = self.get_link_index(link_name)
+		if index:
+			center = np.asarray(p.getLinkState(self.body_id, self.links[index]['id'])[0])
+		return center
+
+	def get_link_index(self, link_name):
+		index = None
+		for i, link in enumerate(self.links):
+			if link['name'] == link_name:
+				index = i
+				break
+		return index
+	
+	def move(self, case):
+		self.movements.simple_move(case)
+
+
+class Movements():
+	def __init__(self, operator, link = b'human/left_hand'):
+		"""
+		This class defines 2 types of movements:
+		-  Re-definable inverse kinematic ones (one_end_effector(), two_end_effectors())
+		-  Harcoded simple ones (simple_move() --> 4 predefined movements)
+		"""
+
+		self.op = operator
+		self.body_id = operator.body_id
+		self.link = link
+
+
+	def get_EF_id(self, link_name):
+		"""
+		Gets the id of the endEffector and the list of free revolute joints 
+		"""
+		freeJoints = []
+		numJoints = p.getNumJoints(self.body_id)
+
+		for i in range(numJoints):
+		    info = p.getJointInfo(self.body_id,i)
+		    if info[2] == p.JOINT_REVOLUTE:
+		        freeJoints.append(i)
+		    if info[12] == link_name:
+		        endEffectorId = i
+
+		return endEffectorId, freeJoints
+
+
+	def apply_chest_and_neck_constraints(self):
+		"""
+		Keeps the chest and neck in a straight position.
+		"""
+		# WHY USE THIS: The inverse kinematics functions use the spine link as often as the other
+		# arm links. However, ergonomically, humans dont move the spine that much. Constraining
+		# the spine and neck recreates a more human-like movement in certain cases.
+
+		# Constraint on the Chest
+		p.createConstraint(self.body_id, self.op.get_link_index('human/spine_1'), -1,
+				   self.op.get_link_index('human/spine_2'), p.JOINT_FIXED,
+				   [0, 0, 0], [0, 0, 0], [0, 0, 1])
+
+		# Constraint on the Neck
+		p.createConstraint(self.body_id, self.op.get_link_index('human/neck'), -1,
+				   self.op.get_link_index('human/spine_2'), p.JOINT_FIXED,
+				   [0, 0, 0], [0, 0, 0], [0, 0, 1])
+
+
+	def one_end_effector(self, pos, maxIter = 100, chest_constraint = True):
+		"""
+		Uses Inverse Kinematics to compute the JointPoses for the position/trajectory of the 
```

Comment:
it looks like it's position only. if it was a trajectory, you'd accept an array of vec3